README.md
This software is in the public domain (there is no license).

~~http://martin.horcicka.eu/python/lock_file/~~


# About this Fork

Since I could not find an active upstream repository ~~nor a working
maintainer contact~~ of
[this pip project](https://pypi.python.org/pypi/lock_file)
I published my small patch here. *Update: the changes might be integrated*
*soon since the maintainer is still active! :)*


# Install

```bash
pip install --user git+https://github.com/ax3l/lock_file.git@master
```


# Documentation

From [here](https://pythonhosted.org/lock_file/):

Lock file is a traditional means of synchronization among processes. In this
module it is implemented as an empty regular file exclusively locked using
`fcntl.lockf`. When it is to be released it is removed by default. However,
if all cooperating processes turn off the removal, they get a guaranteed order
of acquisitions and better scalability.

### Class `LockError`

**Parent class:** `Exception`

Raised when a lock file acquisition is unsuccessful because the lock file is
held by another process.

- `__init__(self, message)`
  Initializes the exception.

- `__repr__(self)`
  Returns the official string representation.

### Class `LockFile`

**Parent class:** `object`

This class represents an acquired lock file. After its releasing most methods
lose their sense and raise a ValueError.

- `__init__(self, path, wait = False, remove = True)`
  Creates and locks the specified file. The wait argument can be set to `True` to
  wait until the lock file can be acquired. The remove argument can be set to
  `False` to keep the file after releasing.
  Raises LockError if the wait argument is `False` and the lock file is held by
  another process. Raises `OSError` or `IOError` if any other error occurs. In
  particular, raises `IOError` with the `errno` attribute set to `errno.EINTR`,
  if waiting for the acquisition is interrupted by a signal.

- `__enter__(self)`
  Returns self.
  Raises `ValueError` if the lock file is released.

- `__exit__(self, exc_type, exc_value, traceback)`
  Calls release.
  Raises exceptions raised by release.

- `__repr__(self)`
  Returns the official string representation.

- `fileno(self)`
  Returns the file descriptor used by the lock file.
  Raises `ValueError` if the lock file is released.

- `get_path(self)`
    Returns the path of the lock file.
    Raises `ValueError` if the lock file is released.

- `release(self)`
    Removes (optionally but default) and closes the lock file.
    Raises ValueError if the lock file is already released.
    Raises `OSError` if any other error occurs.

## Examples

Checking if at most one instance of the script is running:

```python
import sys
from lock_file import LockFile, LockError

try:
    lock_f = LockFile('/var/run/app.lock')
except LockError:
    sys.exit('The script is already running')

try:
    do_something_useful()
finally:
    lock_f.release()
```

Waiting for the lock file acquisition:

```python
from lock_file import LockFile

lock_f = LockFile('/var/run/app.lock', wait = True)
try:
    do_something_useful()
finally:
    lock_f.release()
```

Waiting for the lock file acquisition (in Python 2.5 and higher):

```python
from __future__ import with_statement
from lock_file import LockFile

with LockFile('/var/run/app.lock', wait = True):
    do_something_useful()
```

Waiting for a lock of a [h5py](http://www.h5py.org/) file and writing savely
from various processes in parallel:
```python
import h5py as h5

h5_file_name = "example.h5"

lock_f = LockFile(h5_file_name, wait = True, remove = False)
try:
    h5_f = h5.File(h5_file_name, 'a')
    # do something with h5_f object
    h5_f.close()
finally:
    lock_f.release()
```


CHANGES.txt
# $Id: CHANGES.txt 18 2007-10-24 04:47:12Z horcicka $

Version 2.0:
    Released: 2007-10-24

    New features:
        Use of fcntl.flock has been replaced by fcntl.lockf for better
        portability.

        Removing the lock file before its releasing has been made optional.
        The required behavior can now be set at the LockFile instance creation
        time.

    Fixed bugs:
        Lock file creation mode has been changed to 0666. Previously it was
        0777.

        Unexpected exceptions from os.stat are not being hidden anymore.


Version 1.0:
    Released: 2007-05-28


lock_file.py
# $Id: lock_file.py 18 2007-10-24 04:47:12Z horcicka $

'''Lock file manipulation.

Lock file is a traditional means of synchronization among processes. In
this module it is implemented as an empty regular file exclusively
locked using fcntl.lockf. When it is to be released it is removed by
default. However, if all cooperating processes turn off the removal,
they get a guaranteed order of acquisitions and better scalability.

Example: Checking if at most one instance of the script is running

    import sys
    from lock_file import LockFile, LockError

    try:
        lock_f = LockFile('/var/run/app.lock')
    except LockError:
        sys.exit('The script is already running')

    try:
        do_something_useful()
    finally:
        lock_f.release()

Example: Waiting for the lock file acquisition

    from lock_file import LockFile

    lock_f = LockFile('/var/run/app.lock', wait = True)
    try:
        do_something_useful()
    finally:
        lock_f.release()

Example: Waiting for the lock file acquisition (in Python 2.5 and
higher)

    from __future__ import with_statement
    from lock_file import LockFile

    with LockFile('/var/run/app.lock', wait = True):
        do_something_useful()
'''

__all__ = 'LockError', 'LockFile'

import errno
import fcntl
import os


class LockError(Exception):
    '''Lock error.

    Raised when a lock file acquisition is unsuccessful because the lock
    file is held by another process.
    '''

    def __init__(self, message):
        Exception.__init__(self, message)


    def __repr__(self):
        return self.__class__.__name__ + '(' + repr(self.args[0]) + ')'


class LockFile(object):
    '''Lock file.

    This class represents an acquired lock file. After its releasing
    most methods lose their sense and raise a ValueError.
    '''

    def __init__(self, path, wait = False, remove = True):
        '''Initialize and acquire the lock file.

        Creates and locks the specified file. The wait argument can be
        set to True to wait until the lock file can be acquired. The
        remove argument can be set to False to keep the file after
        releasing.

        Raises LockError if the wait argument is False and the lock file
        is held by another process. Raises OSError or IOError if any
        other error occurs. In particular, raises IOError with the errno
        attribute set to errno.EINTR, if waiting for the acquisition is
        interrupted by a signal.
        '''

        object.__init__(self)
        self._path = path
        self._fd = None
        self._remove = remove

        # Acquire the lock file
        while self._fd is None:
            # Open the file
            fd = os.open(path, os.O_CREAT | os.O_WRONLY, 0o666)
            try:
                # Acquire an exclusive lock
                if wait:
                    fcntl.lockf(fd, fcntl.LOCK_EX)
                else:
                    try:
                        fcntl.lockf(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
                    except IOError as e:
                        if e.errno in (errno.EACCES, errno.EAGAIN):
                            raise LockError(
                                'Lock file is held by another process: '
                                + repr(self._path))
                        else:
                            raise

                # Check if the locked file is the required one (it could
                # have been removed and possibly recreated between the
                # opening and the lock acquisition)
                try:
                    stat1 = os.stat(path)
                except OSError as e:
                    if e.errno != errno.ENOENT:
                        raise
                else:
                    stat2 = os.fstat(fd)
                    if stat1.st_dev == stat2.st_dev \
                        and stat1.st_ino == stat2.st_ino:

                        self._fd = fd

            finally:
                # Close the file if it is not the required one
                if self._fd is None:
                    os.close(fd)


    def __enter__(self):
        if self._fd is None:
            raise ValueError('The lock file is released')

        return self


    def __repr__(self):
        repr_str = '<'
        if self._fd is None:
            repr_str += 'released'
        else:
            repr_str += 'acquired'

        repr_str += ' lock file ' + repr(self._path) + '>'
        return repr_str


    def get_path(self):
        if self._fd is None:
            raise ValueError('The lock file is released')

        return self._path


    def fileno(self):
        if self._fd is None:
            raise ValueError('The lock file is released')

        return self._fd


    def release(self):
        '''Release the lock file.

        Removes (optionally) and closes the lock file.

        Raises ValueError if the lock file is already released. Raises
        OSError if any other error occurs.
        '''

        if self._fd is None:
            raise ValueError('The lock file is already released')

        # Remove and close the file
        try:
            if self._remove:
                os.remove(self._path)
        finally:
            try:
                os.close(self._fd)
            finally:
                self._fd = None


    def __exit__(self, exc_type, exc_value, traceback):
        self.release()
